src/sdk/pynni/nni/compression/torch/pruning/amc/channel_pruning_env.py

```diff
@@ -0,0 +1,497 @@
+                m = m.module
+                if type(m) == nn.Conv2d and m.groups == m.in_channels:  # depth-wise conv, buffer
+                    this_buffer_list.append(i)
+                else:  # really prunable
+                    self.prunable_idx.append(i)
+                    self.prunable_ops.append(m)
+                    self.layer_type_dict[i] = type(m)
+                    self.buffer_dict[i] = this_buffer_list
+                    this_buffer_list = []  # empty
+                    self.org_channels.append(m.in_channels if type(m) == nn.Conv2d else m.in_features)
+
+                    self.strategy_dict[i] = [self.lbound, self.lbound]
+
+        self.strategy_dict[self.prunable_idx[0]][0] = 1  # modify the input
+        self.strategy_dict[self.prunable_idx[-1]][1] = 1  # modify the output
+
+        self.shared_idx = []
+        if self.args.model_type == 'mobilenetv2':  # TODO: to be tested! Share index for residual connection
+            connected_idx = [4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32]  # to be partitioned
+            last_ch = -1
+            share_group = None
+            for c_idx in connected_idx:
+                if self.prunable_ops[c_idx].in_channels != last_ch:  # new group
+                    last_ch = self.prunable_ops[c_idx].in_channels
+                    if share_group is not None:
+                        self.shared_idx.append(share_group)
+                    share_group = [c_idx]
+                else:  # same group
+                    share_group.append(c_idx)
+            self.shared_idx.append(share_group)
+            print('=> Conv layers to share channels: {}'.format(self.shared_idx))
+
+        self.min_strategy_dict = copy.deepcopy(self.strategy_dict)
+
+        self.buffer_idx = []
+        for _, v in self.buffer_dict.items():
+            self.buffer_idx += v
+
+        print('=> Prunable layer idx: {}'.format(self.prunable_idx))
+        print('=> Buffer layer idx: {}'.format(self.buffer_idx))
+        print('=> Initial min strategy dict: {}'.format(self.min_strategy_dict))
+
+        # added for supporting residual connections during pruning
+        self.visited = [False] * len(self.prunable_idx)
+        self.index_buffer = {}
+
+    def _extract_layer_information(self):
+        m_list = list(self.model.modules())
+
+        self.data_saver = []
+        self.layer_info_dict = dict()
+        self.wsize_list = []
+        self.flops_list = []
+
+        from .lib.utils import measure_layer_for_pruning
+
+        # extend the forward fn to record layer info
+        def new_forward(m):
+            def lambda_forward(x):
+                m.input_feat = x.clone()
+                measure_layer_for_pruning(m, x)
+                y = m.old_forward(x)
+                m.output_feat = y.clone()
+                return y
+
+            return lambda_forward
+
+        device = None
+        for idx in self.prunable_idx + self.buffer_idx:  # get all
```

Comment:
I understand for all the prunable layers, you update their forward with new ones. but what is the meaning of `buffer`? why we need to change forward for it?

Reply:
docstring added to class ChannelPruningEnv to explain this.